tachyon/math/polynomials/univariate/radix2_evaluation_domain.h

```diff
@@ -73,6 +85,67 @@ class Radix2EvaluationDomain : public UnivariateEvaluationDomain<F, MaxDegree> {
     return base::bits::SafeLog2Ceiling(num_coeffs) <= F::Config::kTwoAdicity;
   }
 
+  void FFTBatch(RowMajorMatrix<F>& mat) {
+    CHECK_EQ(this->size_, mat.rows());
+    size_t log_h = this->log_size_of_group_;
+    mid_ = log_h / 2;
+
+    // The first half looks like a normal DIT.
+    ReverseMatrixIndexBits(mat);
+    RunParallelRowChunks(mat, roots_vec_[0]);
+
+    // For the second half, we flip the DIT, working in bit-reversed order.
+    ReverseMatrixIndexBits(mat);
+    RunParallelRowChunksReversed(mat, roots_vec_[1]);
+    ReverseMatrixIndexBits(mat);
+  }
+
+  CONSTEXPR_IF_NOT_OPENMP void CosetLDEBatch(RowMajorMatrix<F>& mat,
+                                             size_t added_bits) {
+    CHECK_EQ(this->size_, mat.rows());
+    size_t log_h = this->log_size_of_group_;
+    mid_ = log_h / 2;
+
+    // The first half looks like a normal DIT.
+    ReverseMatrixIndexBits(mat);
+    RunParallelRowChunks(mat, inv_roots_vec_[0]);
+
+    // For the second half, we flip the DIT, working in bit-reversed order.
+    ReverseMatrixIndexBits(mat);
+    RunParallelRowChunksReversed(mat, inv_roots_vec_[1]);
+    // We skip the final bit-reversal, since the next FFT expects bit-reversed
+    // input.
+
+    // Rescale coefficients in two ways:
+    // - divide by number of rows (since we're doing an inverse DFT)
+    // - multiply by powers of the coset shift (see default coset LDE impl for
+    // an explanation)
+    std::vector<F> weights = F::GetSuccessivePowers(
```

Comment:
I really want to cache this, but let's do this later after we implement all the features. You can resolve the comment once you read.